src/cdk/tree/tree.ts

```diff
@@ -216,12 +350,129 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
       dataStream = observableOf(this._dataSource);
     }
 
-    if (dataStream) {
-      this._dataSubscription = dataStream
-        .pipe(takeUntil(this._onDestroy))
-        .subscribe(data => this.renderNodeChanges(data));
-    } else if (typeof ngDevMode === 'undefined' || ngDevMode) {
-      throw getTreeNoValidDataSourceError();
+    if (!dataStream) {
+      if (typeof ngDevMode === 'undefined' || ngDevMode) {
+        throw getTreeNoValidDataSourceError();
+      }
+      return;
+    }
+
+    let expansionModel;
+    if (!this.treeControl) {
+      this._expansionModel = new SelectionModel<K>(true);
+      expansionModel = this._expansionModel;
+    } else {
+      expansionModel = this.treeControl.expansionModel;
+    }
+
+    this._dataSubscription = combineLatest([
+      dataStream,
+      this._nodeType,
+      // NB: the data is unused below, however we add it here to essentially
+      // trigger data rendering when expansion changes occur.
+      expansionModel.changed.pipe(
+        startWith(null),
+        tap(expansionChanges => {
+          this._emitExpansionChanges(expansionChanges);
+        }),
+      ),
+    ])
+      .pipe(
+        switchMap(([data, nodeType]) => {
+          if (nodeType === null) {
+            return observableOf([{renderNodes: data}, nodeType] as const);
+          }
+
+          // If we're here, then we know what our node type is, and therefore can
+          // perform our usual rendering pipeline, which necessitates converting the data
+          return this._convertData(data, nodeType).pipe(
+            map(convertedData => [convertedData, nodeType] as const),
+          );
+        }),
+        takeUntil(this._onDestroy),
+      )
+      .subscribe(([data, nodeType]) => {
+        this._renderDataChanges({nodeType, ...data} as RenderingData<T>);
+      });
```

Comment:
It would be a bit clearer if we could split the creation of this Observable (into its own method) from its subscription here